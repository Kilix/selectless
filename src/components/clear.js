import React from 'react'

import controller from '../controller'

class Clear extends React.Component {
  render() {
    const {clearValue, label, render, ...props} = this.props

    return typeof render === 'undefined'
      ? <span {...props} onClick={clearValue}>
          {label || 'Clear'}
        </span>
      : render({clearValue, label})
  }
}

const enhance = controller(['clearValue'])
export default enhance(Clear)
