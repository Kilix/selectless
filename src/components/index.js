import Item from './item'
import Label from './label'
import List from './list'
import Search from './search'
import Clear from './clear'
import TagList from './tagList'
import Tag from './tag'

export {Item, Label, List, Search, Clear, TagList, Tag}
