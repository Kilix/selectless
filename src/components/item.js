import React from 'react'

import controller from '../controller'

class Item extends React.Component {
  render() {
    const {
      currentRef,
      data,
      isCurrent,
      isSelected,
      onSelectValue,
      render,
      passThrough,
      ...props
    } = this.props
    return typeof render === 'undefined'
      ? <div onClick={() => onSelectValue(data)} ref={currentRef} {...props}>
          {data.label}
        </div>
      : render({data, isCurrent, isSelected, onSelectValue})
  }
}
const enhance = controller(['onSelectValue'])
export default enhance(Item)
