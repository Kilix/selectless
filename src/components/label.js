/* eslint-disable */
import React from 'react';

import controller from '../controller';
import { renderOrCloneComponent } from '../utils';

class Label extends React.Component {
  render() {
    const {
      opened,
      placeholder,
      render,
      selectedValue,
      toggleSelect,
      ...props
    } = this.props;
    const value = selectedValue && selectedValue[0];
    return typeof render === 'undefined'
      ? <div onClick={() => toggleSelect()} {...props}>
          {value ? value.label : placeholder}
        </div>
      : render({ opened, placeholder, toggleSelect, value });
  }
}
const enhance = controller([
  'opened',
  'placeholder',
  'selectedValue',
  'toggleSelect',
]);
export default enhance(Label);
