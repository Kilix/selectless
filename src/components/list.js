import React from 'react'

import omit from 'ramda/src/omit'
import map from 'ramda/src/map'
import findIndex from 'ramda/src/findIndex'
import addIndex from 'ramda/src/addIndex'
import equals from 'ramda/src/equals'
import contains from 'ramda/src/contains'
import compose from 'ramda/src/compose'

import controller from '../controller'
import {renderOrCloneComponent, withKeyboardEvent} from '../utils'

const indexedMap = addIndex(map)

class List extends React.Component {
  render() {
    const {
      currentValue,
      opened,
      options,
      render,
      renderItem,
      selectedValue,
      setRef,
      ...props
    } = this.props
    const myprops = omit(
      [
        'renderItem',
        'caseSensitiveSearch',
        'clearSearchValue',
        'hasSearch',
        'onSelectValue',
        'toggleSelect',
        'toggleSearch',
        'searchValue',
        'selectedValue',
        'setRef',
        'options',
        'passThrough',
      ],
      props
    )

    const items = indexedMap((o, idx) => {
      const isCurrent = currentValue === findIndex(equals(o), options)
      return renderOrCloneComponent(renderItem, {
        key: idx,
        data: o,
        isCurrent,
        isSelected: contains(o, selectedValue),
        passThrough: ['data', 'isCurrent', 'isSelected'],
      })
    }, options)

    return typeof render === 'undefined'
      ? opened
        ? <div {...myprops} role="listbox" ref={ref => setRef(ref)}>
            {items}
          </div>
        : null
      : render({opened, items, setRef})
  }
}

const enhance = compose(
  controller([
    'caseSensitiveSearch',
    'hasSearch',
    'opened',
    'options',
    'searchValue',
    'toggleSelect',
    'toggleSearch',
    'selectedValue',
    'onSelectValue',
  ]),
  withKeyboardEvent
)
export default enhance(List)
