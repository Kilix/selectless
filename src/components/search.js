import React from 'react'
import omit from 'ramda/src/omit'

import controller from '../controller'

class Search extends React.Component {
  state = { value: '' }
  componentWillMount () {
    this.props.toggleSearch(true)
  }
  componentWillReceiveProps (nextProps) {
    if (nextProps.searchValue !== this.state.value) {
      this.setState({ value: nextProps.searchValue })
    }
  }
  onChange = e => {
    const v = e.target.value
    const {
      caseSensitive = false,
      onChangeSearchValue,
      toggleCaseSensitive,
      toggleSelect
    } = this.props
    this.setState({ value: v })
    toggleSelect(true)
    toggleCaseSensitive(caseSensitive)
    onChangeSearchValue(v)
  }
  render () {
    const props = omit(
      [
        'caseSensitive',
        'clearSearchValue',
        'clearValue',
        'onChangeSearchValue',
        'toggleCaseSensitive',
        'toggleSearch',
        'toggleSelect',
        'render',
        'searchValue'
      ],
      this.props
    )
    const {
      caseSensitive,
      clearSearchValue,
      clearValue,
      render,
      searchValue,
      toggleSearch,
      toggleSelect
    } = this.props
    const { value } = this.state
    const ElProps = {
      value,
      onChange: this.onChange,
      onFocus: () => toggleSelect(true)
    }

    return typeof render !== 'undefined'
      ? render({
        caseSensitive,
        clearSearchValue,
        clearValue,
        onChange: this.onChange,
        searchValue,
        toggleSearch,
        toggleSelect,
        value
      })
      : <input type='text' role='combobox' {...props} {...ElProps} />
  }
}

const enhance = controller([
  'clearSearchValue',
  'clearValue',
  'onChangeSearchValue',
  'searchValue',
  'toggleCaseSensitive',
  'toggleSearch',
  'toggleSelect'
])

export default enhance(Search)
