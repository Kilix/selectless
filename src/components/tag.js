/* eslint-disable */
import React from 'react'
import PropTypes from 'prop-types'

import controller from '../controller'
import {renderOrCloneComponent} from '../utils'

class Tag extends React.Component {
  clearTag = e => {
    const {clearOneValue, tag} = this.props
    clearOneValue(tag)
    e.stopPropagation()
  }
  render() {
    const {render, tag} = this.props
    return typeof render === 'undefined'
      ? <span onClick={this.clearTag}>{tag.label}</span>
      : render({tag, clear: this.clearTag})
  }
}

const enhance = controller(['clearOneValue'])
export default enhance(Tag)
