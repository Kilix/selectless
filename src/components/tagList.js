/* eslint-disable */
import React from 'react'
import PropTypes from 'prop-types'
import map from 'ramda/src/map'

import controller from '../controller'
import { renderOrCloneComponent } from '../utils'

class TagList extends React.Component {
  renderTags = map(tag =>
    renderOrCloneComponent(this.props.renderTag, {
      key: `tag_${tag.value}`,
      tag
    })
  )

  render() {
    const {
      toggleSelect,
      placeholder,
      selectedValue,
      render,
      renderTag,
      ...props
    } = this.props

    const ElProps = { ...props, onClick: () => toggleSelect() }
    const tags = this.renderTags(selectedValue)
    return typeof render === 'undefined'
      ? <div {...ElProps}>
          {selectedValue.length === 0 ? placeholder : tags}
        </div>
      : render({
          toggleSelect,
          placeholder,
          tags
        })
  }
}
const enhance = controller(['toggleSelect', 'placeholder', 'selectedValue'])
export default enhance(TagList)
