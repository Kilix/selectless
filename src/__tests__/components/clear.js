/* global it, expect, jest */

import React from 'react'
import {mount} from 'enzyme'
import toJson from 'enzyme-to-json'

import {Clear} from '../../components/'

it('Clear', () => {
  const fn = jest.fn()
  const ctx = {clearValue: fn}
  const tree = mount(<Clear />, {context: ctx})
  expect(toJson(tree)).toMatchSnapshot()
})
it('Custom Clear', () => {
  const fn = jest.fn()
  const ctx = {clearValue: fn}
  const tree = mount(<Clear render={() => <div />} />, {context: ctx})
  expect(toJson(tree)).toMatchSnapshot()
})
it('has default label', () => {
  const fn = jest.fn()
  const ctx = {clearValue: fn}
  const clear = mount(<Clear />, {context: ctx})
  expect(clear.html()).toMatchSnapshot('<span>Clear</span>')
})
it('Simulate click', () => {
  const fn = jest.fn()
  const ctx = {clearValue: fn}
  const tree = mount(<Clear />, {context: ctx})
  tree.simulate('click')
  expect(fn.mock.calls.length).toBe(1)
})
