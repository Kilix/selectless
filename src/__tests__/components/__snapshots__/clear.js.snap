// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<span>Clear</span> 1`] = `"<span>Clear</span>"`;

exports[`Clear 1`] = `
<getContext(Clear)>
  <Clear
    clearValue={[Function]}
  >
    <span
      onClick={[Function]}
    >
      Clear
    </span>
  </Clear>
</getContext(Clear)>
`;

exports[`Custom Clear 1`] = `
<getContext(Clear)
  render={[Function]}
>
  <Clear
    clearValue={[Function]}
    render={[Function]}
  >
    <div />
  </Clear>
</getContext(Clear)>
`;
